### PR TITLE
refactor(wslserviceapi): Remove double server in WSL

### PR DESCRIPTION
**Pull request by EduardGomezEscandell**
New diagram 🖌️ 

image

---

UDENG-2509

### CODE REVIEW

**Review by didrocks — CHANGES_REQUESTED** [review]
Really happy to see this!

So, there are many comments, but nothing on the logic itself. Mostly formatting and for better maintainability in the long term.
Thanks again for tackling this non trivial tasks!

Btw, you need to update your diagram :)

File: windows-agent/internal/proservices/wslinstance/proattachement.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	if err := client.WaitReady(ctx); err != nil {
+		return err
+	}
+
+	// Block until the connection drops
+	client.WaitDone(ctx)
+	return nil
+}
+
+// SetProStream sets the pro attachment stream for the client.
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply windows-agent/internal/proservices/wslinstance/proattachement.go:36]: Where is golangci-lint when you need it 😭

File: windows-agent/internal/tasks/pro_attachment.go [hunk 2]
```diff
@@ -22,8 +21,8 @@ type ProAttachment struct {
 }
 
 // Execute is needed to fulfil Task.
-func (t ProAttachment) Execute(ctx context.Context, client wslserviceapi.WSLClient) error {
-	_, err := client.ApplyProToken(ctx, &wslserviceapi.ProAttachInfo{Token: t.Token})
+func (t ProAttachment) Execute(ctx context.Context, conn task.Connection) error {
+	err := conn.SendProAttachment(t.Token)
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply windows-agent/internal/tasks/pro_attachment.go:25]: Yess

File: wsl-pro-service/internal/streams/multiclient.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// It only provides communication primitives, it does not handle the logic of the messages themselves.
+type multiClient struct {
+	mainStream agentapi.WSLInstance_ConnectedClient
+	proStream  agentapi.WSLInstance_ProAttachmentCommandsClient
+	lpeStream  agentapi.WSLInstance_LandscapeConfigCommandsClient
+}
+
+// Connect connects to the three streams. Call Close to release resources.
+//
+//nolint:revive // This method is only public to tests, where multiClient has an alias available
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply wsl-pro-service/internal/streams/multiclient.go:22]: That leads to another false positive in the export_test file 😄 

```go
// export_test.go

type MultiClient = multiClient

func Connect(ctx context.Context, conn *grpc.ClientConn) (c *MultiClient, err error) {
//                                                           ^^^^^^^^^^^
// unexported-return: exported func Connect returns unexported type *streams.multiClient, which can be annoying to use (revive)
	return connect(ctx, conn)
}
```

File: wsl-pro-service/internal/streams/server.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
+// Package streams abstracts the bi-directional gRPC stream and provides a faux server that mimics a unary call server.
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply wsl-pro-service/internal/streams/server.go:1]: Fake: ❌ cheap copy, untrustworthy
Faux: 🥖 Refined, elegant disguise

File: wsl-pro-service/internal/streams/server.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+	for {
+		// Graceful stop
+		select {
+		case <-gCtx.Done():
+			return nil
+		default:
+		}
+
+		log.Debugf(ctx, "Started serving %s requests", reflect.TypeOf((*Command)(nil)).Elem())
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply wsl-pro-service/internal/streams/server.go:197]: Clearly!

File: windows-agent/internal/proservices/wslinstance/client.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+// SetConnectedStream sets the Connected stream for the client.
+// This step is necessary for WaitReady to return.
+func (c *client) SetConnectedStream(stream agentapi.WSLInstance_ConnectedServer) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.connStream != nil {
+		return errors.New("stream already connected")
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply windows-agent/internal/proservices/wslinstance/client.go:128]: Fun fact: there was a bug here! Kinda, depends on what the desired behaviour is. Connecting with a secnd stream would drop both the first and second ones.

File: windows-agent/internal/proservices/wslinstance/wslinstance.go [hunk 7]
```diff
@@ -28,52 +26,59 @@ type Service struct {
...
+	client, info, err := mainHandshake(ctx, s, stream.Recv)
 	if err != nil {
-		return fmt.Errorf("WSLInstance service: incomplete handshake: did not receive info from WSL distro: %v", err)
+		return err
 	}
+	defer client.Close()
 
-	distroName := info.GetWslName()
+	if err := client.SetConnectedStream(stream); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply windows-agent/internal/proservices/wslinstance/wslinstance.go:56]: A single test can cover all three cases :)

File: windows-agent/internal/proservices/wslinstance/client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (didrocks) [reply windows-agent/internal/proservices/wslinstance/client.go:122]: \o/

File: wsl-pro-service/internal/streams/server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (didrocks) [reply wsl-pro-service/internal/streams/server.go:1]: De toute évidence !

File: windows-agent/internal/proservices/wslinstance/proattachment.go
Comment (CarlosNihelton) [inline windows-agent/internal/proservices/wslinstance/proattachment.go:1]: 🤣

**Review by didrocks — CHANGES_REQUESTED** [review]
Ok, there is only one comment about the exported function (despite the other false positive linting warning) that I think is worth changing! Thanks for fixing them all so quickly :)

File: windows-agent/internal/proservices/wslinstance/client.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+	c.service.clientsMu.Lock()
+	defer c.service.clientsMu.Unlock()
+
+	delete(c.service.clients, c.name)
+}
+
+// msgToError translates a result received via gRPC into an error.
+// If there is a problem translating, an error will be returned and the second argument
+// will be false.
```
Comment (CarlosNihelton) [inline windows-agent/internal/proservices/wslinstance/client.go:97]:
```suggestion
// msgToError translates a result received via gRPC into an error.
// If there is a problem translating, an error will be returned and the first return value
// will be false.
```

File: windows-agent/internal/proservices/wslinstance/landscapeconfig.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	if err := client.WaitReady(ctx); err != nil {
+		return err
+	}
+
+	// Block until the connection drops
+	client.WaitDone(ctx)
+	return nil
+}
+
+// SetLandscapeConfigStream sets the landscape config stream for the client.
```
Comment (CarlosNihelton) [inline windows-agent/internal/proservices/wslinstance/landscapeconfig.go:35]:
```suggestion
// SetLandscapeConfigStream sets the Landscape config stream for the client.
```

I think we should capitalize all instances of the word `Landscape`.

**Review by CarlosNihelton — COMMENTED** [review]
Two nitpicks for you.

I must rely on both of you guys on the implementation details. This is far too big to fit in my head, despite its considerable size :)

**Comment by EduardGomezEscandell** [comment]
Yeah we did a pair review with liveshare and all, too big a PR to review the usual way :)

**Review by didrocks — APPROVED** [review]
Thanks for the overall logical simplifcation!

**Pull request merged**